pkg/controller/node/cidr_allocator.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/big"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/util"
+)
+
+const (
+	// cidrsBufSize is the buffer size of the free CIDRs channel
+	cidrsBufSize = 16
+	// nextFreeCIDRCooldown is the cooldown time before a released CIDR could be reused
+	nextFreeCIDRCooldown = 500 * time.Millisecond
+	// allocateNextTimeout control how long AllocateNext() will block before it return an error
+	allocateNextTimeout = 1 * time.Second
+)
+
+var errCIDRRangeNoCIDRsRemaining = errors.New("CIDR allocation failed; there are no remaining CIDRs left to allocate in the accepted range")
+
+// CIDRAllocator is an interface implemented by things that know how to allocate/recycle CIDR for nodes.
+type CIDRAllocator interface {
+	AllocateNext() (*net.IPNet, error)
+	Release(*net.IPNet)
+}
+
+// emptyAllocator is a CIDRAllocator implementation that allocates empty CIDRs, yielding
+// a CIDR value of nil for every call to AllocateNext().
+type emptyAllocator struct{}
+
+// AllocateNext always returns nil and error
+func (r *emptyAllocator) AllocateNext() (cidr *net.IPNet, err error) {
+	return nil, errCIDRRangeNoCIDRsRemaining
+}
+
+// Release is a noop
+func (r *emptyAllocator) Release(_ *net.IPNet) {
+	// noop
+}
+
+type rangeAllocator struct {
+	clusterCIDR    *net.IPNet
+	subNetMaskSize int
+	cidrs          chan *net.IPNet
+	used           big.Int
+	lock           sync.Mutex
+	rand           *rand.Rand
+}
+
+// NewCIDRRangeAllocator returns a CIDRAllocator to allocate CIDR for node
+// If subNetMaskSize is less than cluster CIDR mask size, an emptyAllocator is returned.
+// Otherwise an rangeAllocator is returned.
+func NewCIDRRangeAllocator(clusterCIDR *net.IPNet, subNetMaskSize int) CIDRAllocator {
+	cidrMask := clusterCIDR.Mask
+	cidrMaskSize, _ := cidrMask.Size()
+
+	if subNetMaskSize < cidrMaskSize {
+		return &emptyAllocator{}
```

Comment:
Instead of successfully returning an emptyAllocator, could we make this New() function return an error? We know up front if the allocator has been created with impossible parameters. It feels silly to return something that we know will fail whenever it first gets used.